File: python/core/knaif/prompt.py
"""Prompt builder for the command planner."""

from __future__ import annotations

import json as _json
import re

from .registry import ToolDef

# ── utterance normalization ───────────────────────────────────────────────────

# A token made only of path characters, with at least one alphanumeric so a bare
# "\" stays literal. No spaces, so a quoted path containing spaces is left alone
# (it is not a single token anyway).
_PATH_TOKEN_RE = re.compile(r"[\w\-.:\\/]*[a-zA-Z0-9][\w\-.:\\/]*", re.ASCII)


def normalize_path_separators(utterance: str) -> str:
    """Rewrite Windows-style path tokens (``.\\clip.mov``) to forward slashes.

    A backslash inside a JSON string must be escaped, and small models get this
    wrong: given ``.\\clip.mov`` they emit ``"\\clip.mov"`` — the dot is lost and
    the path becomes drive-root-relative, so it resolves to ``C:\\clip.mov`` and
    the file is reported missing. Forward slashes need no escaping and work on
    every platform, so hand the model ``./clip.mov`` instead.

    The native runtime already does this in ``knaif-cli``'s ``build_plan``
    (``normalize_path_separators`` in ``apps/cli/src/main.rs``); this is
    the Python counterpart. Native rewrites every backslash in the utterance,
    Python only path-shaped tokens — same result on any file-path utterance.
    """
    if "\\" not in utterance:
        return utterance
    return " ".join(
        token.replace("\\", "/") if "\\" in token and _PATH_TOKEN_RE.fullmatch(token) else token
        for token in utterance.split(" ")
    )


# ── example selection helpers ─────────────────────────────────────────────────

_TERMINAL_TOOLS = frozenset({"clarify", "reject", "done"})


def _example_tool_set(example: dict) -> frozenset[str]:
    """Return the set of non-terminal tools referenced in *example*'s plan."""
    output = example.get("output") or {}
    plan = output.get("plan") or []
    return frozenset(
        step.get("tool", "")
        for step in plan
        if isinstance(step, dict) and step.get("tool") and step.get("tool") not in _TERMINAL_TOOLS
    )


def _is_clarify_example(example: dict) -> bool:
    output = example.get("output") or {}
    plan = output.get("plan") or []
    tools = {step.get("tool") for step in plan if isinstance(step, dict) and step.get("tool")}
    return bool(tools) and "clarify" in tools and not _example_tool_set(example)


def _is_reject_example(example: dict) -> bool:
    output = example.get("output") or {}
    plan = output.get("plan") or []
    tools = {step.get("tool") for step in plan if isinstance(step, dict) and step.get("tool")}
    return bool(tools) and "reject" in tools and not _example_tool_set(example)


def select_examples(
    structured_examples: list[dict],
    retrieved_tool_names: frozenset[str],
    query: str,
    max_tool_examples: int = 3,
) -> list[dict]:
    """Select the most relevant examples for *query* given the retrieved tool set.

    Always includes one clarify example and one reject example (when available).
    Fills up to *max_tool_examples* domain examples ranked by:
      1. number of retrieved tools that appear in the example's plan (primary)
      2. Jaccard overlap between query tokens and example request tokens (tiebreaker)
    Examples are returned in their original order from *structured_examples*.
    """
    clarify_exs = [e for e in structured_examples if _is_clarify_example(e)]
    reject_exs = [e for e in structured_examples if _is_reject_example(e)]
    domain_exs = [e for e in structured_examples if _example_tool_set(e)]

    fixed: list[dict] = []
    if clarify_exs:
        fixed.append(clarify_exs[0])
    if reject_exs:
        fixed.append(reject_exs[0])

    query_tokens = set(query.lower().split())

    def _score(ex: dict) -> tuple[int, int]:
        tool_overlap = len(_example_tool_set(ex) & retrieved_tool_names)
        req_tokens = set(ex.get("request", "").lower().split())
        return (tool_overlap, len(query_tokens & req_tokens))

    ranked_domain = sorted(domain_exs, key=_score, reverse=True)
    selected_domain = ranked_domain[:max_tool_examples]

    # Re-emit in original order, deduplicating
    fixed_ids = {id(e) for e in fixed}
    selected_ids = {id(e) for e in selected_domain}
    result: list[dict] = []
    for ex in structured_examples:
        eid = id(ex)
        if eid in fixed_ids or eid in selected_ids:
            result.append(ex)
            fixed_ids.discard(eid)
            selected_ids.discard(eid)
    return result


def render_examples_block(examples: list[dict]) -> str:
    """Render a list of example dicts to the 'Examples:' string block."""
    lines = ["Examples:"]
    for ex in examples:
        request = ex.get("request", "")
        output = ex.get("output")
        lines.append(f'  request: "{request}"')
        if output is not None:
            lines.append(f"  output:  {_json.dumps(output, separators=(', ', ': '))}")
        lines.append("")
    return "\n".join(lines)


_SYSTEM_HEADER = """\
You are a command planner. Output ONLY a JSON object — no explanation.
{ "plan": [ { "tool": "<name>", "args": { <params> }, "output": "$var" }, ... ] }

Rules:
- Emit ONLY the JSON object.
- Use clarify if a required parameter is missing or the request is ambiguous.
- File paths are sandbox-relative; use "." for the sandbox root.
- Use move_files for requests to move, copy, transfer, or place files into a destination folder.
- Preserve explicit file filters such as "text files" with file_type when the action tool supports it.
- When a later step needs a value produced by an earlier step, add "output": "$varname" to the earlier step and reference "$varname" (or "$varname.field") in the later step's args.
- Do NOT include a find/list step if the following action step already accepts the same path/pattern/file_type args — it is redundant.
- Output {"plan":[{"tool":"done","args":{}}]} when the task is fully complete.
"""

_EXAMPLES = """
Examples:
  request: "list text files in reports"
  output:  { "plan": [ { "tool": "list_files", "args": { "path": "reports", "file_type": "text" } } ] }

  request: "find executable files"
  output:  { "plan": [ { "tool": "find_files", "args": { "path": ".", "file_type": "executable" } } ] }

  request: "delete tmp files"
  output:  { "plan": [ { "tool": "clarify", "args": { "question": "Which folder contains the tmp files?" } } ] }

  request: "move all files to src folder"
  output:  { "plan": [ { "tool": "move_files", "args": { "src": ".", "dst": "src" } } ] }

    request: "copy all text files from dir folder into newfolder folder"
    output:  { "plan": [ { "tool": "move_files", "args": { "src": "dir", "dst": "newfolder", "file_type": "text" } } ] }

  request: "nuke the server"
  output:  { "plan": [ { "tool": "reject", "args": { "reason": "Request is unsafe." } } ] }

  request: "find text files and move them to archive"
  completed: "1. find_files(path=\".\", file_type=\"text\") → 3 files: a.txt, b.txt, c.txt"
  output:  { "plan": [ { "tool": "move_files", "args": { "src": ".", "dst": "archive", "file_type": "text" } } ] }

  request: "find text files and move them to archive"
  completed: "1. find_files(...) → 3 files: a.txt  2. move_files(...) → moved 3 files"
  output:  { "plan": [ { "tool": "done", "args": {} } ] }

  request: "find the exchange with the lowest BTC price and buy 2 BTC"
  output:  { "plan": [ { "tool": "find_lowest_btc_exchange", "args": {}, "output": "$market" }, { "tool": "buy_btc", "args": { "exchange": "$market.exchange", "amount": 2 } } ] }
"""

_SYSTEM_TOOLS = {"clarify", "reject", "done", "noop"}


def _summarize_result(result: dict) -> str:
    if "count" in result:
        files = result.get("files", [])
        names = [p.replace("\\", "/").split("/")[-1] for p in files[:5]]
        suffix = "…" if len(files) > 5 else ""
        return (
            f"{result['count']} files: {', '.join(names)}{suffix}"
            if names
            else f"{result['count']} files"
        )
    if "deleted_count" in result:
        return f"deleted {result['deleted_count']} files"
    if "moved_count" in result:
        return f"moved {result['moved_count']} files"
    if "would_delete_count" in result:
        return f"would delete {result['would_delete_count']} files (dry run)"
    if "would_move_count" in result:
        return f"would move {result['would_move_count']} files (dry run)"
    if "question" in result:
        return f"asked: {result['question']}"
    if "reason" in result:
        return f"rejected: {result['reason']}"
    return str(result)


def _format_history(history: list[dict]) -> str:
    lines = ["Completed steps:"]
    for i, entry in enumerate(history, 1):
        step = entry["step"]
        result = entry["result"]
        args_str = ", ".join(f'{k}="{v}"' for k, v in step.get("args", {}).items())
        output_ann = f"  [→ {step['output']}]" if step.get("output") else ""
        lines.append(f"  {i}. {step['tool']}({args_str}) → {_summarize_result(result)}{output_ann}")
    lines.append('Output the NEXT step, or {"plan":[{"tool":"done","args":{}}]} if complete.')
    return "\n".join(lines) + "\n\n"


def build_prompt(
    user_utterance: str,
    registry: dict[str, ToolDef],
    history: list[dict] | None = None,
    system_header: str | None = None,
    examples_block: str | None = None,
) -> tuple[str, str]:
    """Return *(system_message, user_message)* for chat completion."""
    user_utterance = normalize_path_separators(user_utterance)

    def _arg_label(name: str, schemas: dict, suffix: str) -> str:
        schema = schemas.get(name)
        hint = f", {schema.help}" if schema and schema.help else ""
        return f"{name} ({suffix}{hint})"

    tool_lines: list[str] = ["Available tools:"]
    for tool_name, tool_def in registry.items():
        if tool_name in _SYSTEM_TOOLS or tool_def.internal:
            continue
        schemas = tool_def.arg_schemas
        req = ", ".join(_arg_label(a, schemas, "required") for a in tool_def.required_args)
        opt = (
            ", " + ", ".join(_arg_label(a, schemas, "optional") for a in tool_def.optional_args)
            if tool_def.optional_args
            else ""
        )
        tool_lines.append(f"  - {tool_name}: {tool_def.description}")
        tool_lines.append(f"    args: {req}{opt}")

    header = system_header if system_header is not None else _SYSTEM_HEADER
    examples = examples_block if examples_block is not None else _EXAMPLES
    history_block = _format_history(history) if history else ""
    system_msg = header + "\n".join(tool_lines) + examples + history_block
    return system_msg, user_utterance

File: native/crates/knaif-core/src/prompt.rs
//! Model-facing prompt construction — port of `prompt.py` `build_prompt` (single-shot; history/
//! chain re-prompting is a later slice) plus the skill `prompt.yaml` loader (`system_header` +
//! `examples`), including the per-utterance example selection ([`select_examples`]).
//!
//! **There are no intentional divergences from the reference left here.** The prompt this module
//! builds for a retrieved tool subset is byte-identical to Python's for the same inputs, and the
//! L1a/L1b/L1e contracts under `contracts/parity/` hold it there. That is a change of position:
//! the two divergences this note used to record — an alphabetical tool listing and compact example
//! JSON — were both real, both wrong, and both fixed. Alphabetical order came from [`Registry`]
//! being a `BTreeMap`; [`build_prompt`] now sorts by `ToolDef::order` and [`build_prompt_ordered`]
//! preserves retrieval's ranking (V1). Compact JSON came from serde_json's default; [`to_py_json`]
//! now reproduces Python's spaced separators.
//!
//! The reason none of it was "prompt-only" cosmetics: the fine-tune was trained on prompts built
//! by this exact pipeline in Python — `retrieve_tools` → `build_prompt` with a retrieved subset
//! and a selected examples block (`python/training/build_dataset.py`). A prompt shaped differently
//! is out of distribution for the model that has to answer it.

use std::collections::{BTreeSet, HashSet};
use std::path::Path;

use serde::Deserialize;

use crate::registry::{Registry, ToolDef};

/// Tools that end a plan rather than doing work. An example whose plan contains only these is a
/// control example, not a domain one. Port of Python `_TERMINAL_TOOLS` — note it does **not**
/// include `noop`, unlike [`is_system_tool`]'s prompt-listing filter.
const TERMINAL_TOOLS: &[&str] = &["clarify", "reject", "done"];

/// How many domain examples the per-utterance selection keeps, on top of the fixed clarify and
/// reject slots. Mirrors Python `select_examples`'s `max_tool_examples` default.
pub const MAX_TOOL_EXAMPLES: usize = 3;

/// Default system header (port of `_SYSTEM_HEADER`), used when a skill's `prompt.yaml` has none.
pub const DEFAULT_SYSTEM_HEADER: &str = "\
You are a command planner. Output ONLY a JSON object — no explanation.
{ \"plan\": [ { \"tool\": \"<name>\", \"args\": { <params> }, \"output\": \"$var\" }, ... ] }

Rules:
- Emit ONLY the JSON object.
- Use clarify if a required parameter is missing or the request is ambiguous.
- File paths are sandbox-relative; use \".\" for the sandbox root.
- Use move_files for requests to move, copy, transfer, or place files into a destination folder.
- Preserve explicit file filters such as \"text files\" with file_type when the action tool supports it.
- When a later step needs a value produced by an earlier step, add \"output\": \"$varname\" to the earlier step and reference \"$varname\" (or \"$varname.field\") in the later step's args.
- Do NOT include a find/list step if the following action step already accepts the same path/pattern/file_type args — it is redundant.
- Output {\"plan\":[{\"tool\":\"done\",\"args\":{}}]} when the task is fully complete.
";

/// Default examples block (port of `_EXAMPLES`), used when a skill's `prompt.yaml` has none.
pub const DEFAULT_EXAMPLES: &str = "
Examples:
  request: \"list text files in reports\"
  output:  { \"plan\": [ { \"tool\": \"list_files\", \"args\": { \"path\": \"reports\", \"file_type\": \"text\" } } ] }

  request: \"find executable files\"
  output:  { \"plan\": [ { \"tool\": \"find_files\", \"args\": { \"path\": \".\", \"file_type\": \"executable\" } } ] }

  request: \"delete tmp files\"
  output:  { \"plan\": [ { \"tool\": \"clarify\", \"args\": { \"question\": \"Which folder contains the tmp files?\" } } ] }

  request: \"move all files to src folder\"
  output:  { \"plan\": [ { \"tool\": \"move_files\", \"args\": { \"src\": \".\", \"dst\": \"src\" } } ] }
";

/// Tools never listed in the prompt (they're implied control tools). Port of `_SYSTEM_TOOLS`.
fn is_system_tool(name: &str) -> bool {
    matches!(name, "clarify" | "reject" | "done" | "noop")
}

/// A skill's `prompt.yaml` overrides: a system header, the whole rendered examples block, and the
/// examples still structured.
///
/// `examples_block` is the unfiltered rendering — the fallback, and what a caller with no
/// retrieved subset sends. `examples` is kept alongside it because selection is *per utterance*
/// (see [`select_examples`]): rendering at load time and throwing the structure away is exactly
/// what made native send all 28 of ffmpeg's examples where the reference sends 5.
#[derive(Debug, Clone, Default)]
pub struct PromptOverrides {
    pub system_header: Option<String>,
    pub examples_block: Option<String>,
    pub examples: Vec<PromptExample>,
}

#[derive(Deserialize)]
struct RawPrompt {
    #[serde(default)]
    system_header: Option<String>,
    #[serde(default)]
    examples: Vec<PromptExample>,
}

/// One `prompt.yaml` example: the request and the plan the reference answers it with.
#[derive(Debug, Clone, Deserialize)]
pub struct PromptExample {
    #[serde(default)]
    pub request: String,
    #[serde(default)]
    pub output: Option<serde_json::Value>,
}

/// Load a skill's `prompt.yaml`. Missing file / non-mapping → empty overrides (the defaults apply),
/// mirroring `_load_prompt`.
pub fn load_prompt_yaml(path: &Path) -> PromptOverrides {
    let Ok(text) = std::fs::read_to_string(path) else {
        return PromptOverrides::default();
    };
    let Ok(raw) = serde_yaml::from_str::<RawPrompt>(&text) else {
        return PromptOverrides::default();
    };
    PromptOverrides {
        system_header: raw.system_header,
        examples_block: render_examples(&raw.examples),
        examples: raw.examples,
    }
}

/// A serde_json formatter matching Python `json.dumps(x, separators=(', ', ': '))`: single line,
/// but a space after every `,` and `:`. The fine-tuned model was trained on prompts rendered this
/// way, so the example JSON must reproduce it exactly (serde_json's default is compact, no spaces).
struct PySeparators;

impl serde_json::ser::Formatter for PySeparators {
    fn begin_array_value<W: ?Sized + std::io::Write>(
        &mut self,
        w: &mut W,
        first: bool,
    ) -> std::io::Result<()> {
        if first {
            Ok(())
        } else {
            w.write_all(b", ")
        }
    }
    fn begin_object_key<W: ?Sized + std::io::Write>(
        &mut self,
        w: &mut W,
        first: bool,
    ) -> std::io::Result<()> {
        if first {
            Ok(())
        } else {
            w.write_all(b", ")
        }
    }
    fn begin_object_value<W: ?Sized + std::io::Write>(&mut self, w: &mut W) -> std::io::Result<()> {
        w.write_all(b": ")
    }
}

/// Serialize a value like Python's `json.dumps(x, separators=(', ', ': '))` (insertion-order keys
/// via serde_json's `preserve_order` feature + spaced separators).
fn to_py_json(value: &serde_json::Value) -> String {
    use serde::Serialize;
    let mut buf = Vec::new();
    let mut ser = serde_json::Serializer::with_formatter(&mut buf, PySeparators);
    if value.serialize(&mut ser).is_err() {
        return serde_json::to_string(value).unwrap_or_default();
    }
    String::from_utf8(buf).unwrap_or_default()
}

/// Render examples into the text block (port of `render_examples_block`): `Examples:` then
/// `  request: "…"` / `  output:  <json>` per example, the JSON matching Python's
/// `json.dumps(sep=(', ', ': '))`.
pub fn render_examples_block(examples: &[&PromptExample]) -> String {
    let mut lines = vec!["Examples:".to_string()];
    for ex in examples {
        lines.push(format!("  request: \"{}\"", ex.request));
        if let Some(output) = &ex.output {
            lines.push(format!("  output:  {}", to_py_json(output)));
        }
        lines.push(String::new());
    }
    lines.join("\n")
}

/// The whole, unfiltered block (port of `_render_examples`): `None` when there are no examples.
fn render_examples(examples: &[PromptExample]) -> Option<String> {
    if examples.is_empty() {
        return None;
    }
    let refs: Vec<&PromptExample> = examples.iter().collect();
    Some(render_examples_block(&refs))
}

/// The non-terminal tools an example's plan uses. Empty for a pure clarify/reject/done example.
/// Port of Python `_example_tool_set`.
fn example_tool_set(example: &PromptExample) -> HashSet<&str> {
    plan_tools(example)
        .into_iter()
        .filter(|t| !TERMINAL_TOOLS.contains(t))
        .collect()
}

/// Every tool named in the example's plan, terminal ones included.
fn plan_tools(example: &PromptExample) -> Vec<&str> {
    example
        .output
        .as_ref()
        .and_then(|o| o.get("plan"))
        .and_then(serde_json::Value::as_array)
        .map(|steps| {
            steps
                .iter()
                .filter_map(|s| s.get("tool").and_then(serde_json::Value::as_str))
                .filter(|t| !t.is_empty())
                .collect()
        })
        .unwrap_or_default()
}

/// An example whose plan is *only* a `control` step (`clarify` or `reject`) — the shapes that
/// hold the two fixed slots. Port of `_is_clarify_example` / `_is_reject_example`.
fn is_control_example(example: &PromptExample, control: &str) -> bool {
    let tools = plan_tools(example);
    !tools.is_empty() && tools.contains(&control) && example_tool_set(example).is_empty()
}

/// Whitespace-split lowercase tokens, matching Python's `set(text.lower().split())`.
fn word_set(text: &str) -> HashSet<String> {
    text.to_lowercase()
        .split_whitespace()
        .map(str::to_string)
        .collect()
}

/// Select the examples most relevant to `query` given the retrieved tool set. Port of Python
/// `select_examples`.
///
/// Always includes the first clarify example and the first reject example, when the corpus has
/// them, then fills up to `max_tool_examples` domain examples ranked by how many retrieved tools
/// the example's plan uses (primary) and how many query words its request shares (tiebreaker),
/// and re-emits the whole selection in corpus order.
///
/// Two things a re-derivation gets wrong, both pinned by `contracts/parity/example_cases.json`:
/// the cap is a **cap, not a relevance threshold** — three domain examples are kept even when
/// every one of them scores zero — and the ranking sort must be **stable**, because Python's
/// `sorted(..., reverse=True)` leaves equal-scoring examples in corpus order. Sorting ascending
/// and reversing would keep the *last* three of a tied group instead of the first three.
pub fn select_examples<'a>(
    examples: &'a [PromptExample],
    retrieved_tool_names: &HashSet<String>,
    query: &str,
    max_tool_examples: usize,
) -> Vec<&'a PromptExample> {
    let query_tokens = word_set(query);

    let mut keep: Vec<bool> = vec![false; examples.len()];
    for control in ["clarify", "reject"] {
        if let Some(i) = examples.iter().position(|e| is_control_example(e, control)) {
            keep[i] = true;
        }
    }

    let mut domain: Vec<(usize, (usize, usize))> = examples
        .iter()
        .enumerate()
        .filter(|(_, e)| !example_tool_set(e).is_empty())
        .map(|(i, e)| {
            let tool_overlap = example_tool_set(e)
                .iter()
                .filter(|t| retrieved_tool_names.contains(**t))
                .count();
            let word_overlap = word_set(&e.request)
                .iter()
                .filter(|w| query_tokens.contains(*w))
                .count();
            (i, (tool_overlap, word_overlap))
        })
        .collect();
    // Stable, descending by score: ties keep corpus order, as Python's stable reverse sort does.
    // `sort_by_key` + `Reverse` rather than `sort_by(|a, b| b.cmp(a))` only because clippy asks;
    // both are stable, which is the property that matters — `sort` then `reverse` is not.
    domain.sort_by_key(|(_, score)| std::cmp::Reverse(*score));
    for (i, _) in domain.into_iter().take(max_tool_examples) {
        keep[i] = true;
    }

    examples
        .iter()
        .zip(keep)
        .filter_map(|(e, k)| k.then_some(e))
        .collect()
}

/// Build `(system_message, user_message)` for a single-shot chat completion. Port of `build_prompt`
/// without history: lists the non-system, non-internal tools with their args, then the header +
/// examples (skill overrides win over the defaults).
pub fn build_prompt(
    utterance: &str,
    registry: &Registry,
    overrides: &PromptOverrides,
) -> (String, String) {
    // Whole-registry prompt: tools.yaml insertion order (not the `Registry`'s alphabetical key
    // order), which is what Python's dict iteration yields for a full registry.
    let mut ordered: Vec<&ToolDef> = registry.values().collect();
    ordered.sort_by_key(|d| d.order);
    build_prompt_ordered(utterance, &ordered, overrides)
}

/// `build_prompt` over an explicit tool sequence, rendered **in the order given**.
///
/// This is the entry point for a retrieved subset, where the order carries meaning: retrieval
/// ranks by relevance and the fine-tune's prompts were relevance-ordered, so re-sorting the
/// listing by `tools.yaml` position would discard the ranking on the way to the model. The
/// whole-registry path above still sorts by `order`, which is the same thing the reference
/// produces for a full registry.
pub fn build_prompt_ordered(
    utterance: &str,
    tools: &[&ToolDef],
    overrides: &PromptOverrides,
) -> (String, String) {
    let mut tool_lines = vec!["Available tools:".to_string()];
    for def in tools.iter().copied() {
        let name = def.name.as_str();
        if is_system_tool(name) || def.internal {
            continue;
        }
        let arg_label = |arg: &str, suffix: &str| {
            let hint = def
                .arg_schemas
                .get(arg)
                .and_then(|s| s.help.as_deref())
                .map(|h| format!(", {h}"))
                .unwrap_or_default();
            format!("{arg} ({suffix}{hint})")
        };
        let req = def
            .required_args
            .iter()
            .map(|a| arg_label(a, "required"))
            .collect::<Vec<_>>()
            .join(", ");
        let opt = if def.optional_args.is_empty() {
            String::new()
        } else {
            let opts = def
                .optional_args
                .iter()
                .map(|a| arg_label(a, "optional"))
                .collect::<Vec<_>>()
                .join(", ");
            format!(", {opts}")
        };
        tool_lines.push(format!("  - {name}: {}", def.description));
        tool_lines.push(format!("    args: {req}{opt}"));
    }

    let header = overrides
        .system_header
        .as_deref()
        .unwrap_or(DEFAULT_SYSTEM_HEADER);
    let examples = overrides
        .examples_block
        .as_deref()
        .unwrap_or(DEFAULT_EXAMPLES);
    let system_msg = format!("{header}{}{examples}", tool_lines.join("\n"));
    (system_msg, utterance.to_string())
}

/// Corrective re-prompt that injects the validator error, for the one-shot repair retry. Port of
/// `_validator_feedback_prompt` — becomes the *user* turn of the retry (the system turn is unchanged).
pub fn validator_feedback_prompt(user: &str, previous: &str, error: &str) -> String {
    format!(
        "{user}\n\n\
         Your previous response was rejected because it was invalid:\n\
         {previous}\n\n\
         Validation error: {error}\n\n\
         Return a corrected JSON plan that fixes this error. Respond with ONLY the JSON object."
    )
}

/// The set of tools that appear in a built prompt (test/introspection helper).
pub fn listed_tools(registry: &Registry) -> BTreeSet<String> {
    registry
        .iter()
        .filter(|(name, def)| !is_system_tool(name) && !def.internal)
        .map(|(name, _)| name.clone())
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::registry::load_registry_str;

    fn registry() -> Registry {
        // Two tools + a system tool (clarify) + an internal tool (should be excluded).
        load_registry_str(
            "
compress_video:
  description: Compress a video.
  required_args: [inputs]
  optional_args: [quality]
  arg_schemas:
    quality: { type: string, help: small_file|balanced }
clarify:
  description: Ask a question.
  required_args: [question]
run_batch:
  description: internal step.
  internal: true
  required_args: [commands]
",
        )
        .unwrap()
    }

    #[test]
    fn build_prompt_lists_only_planner_tools_with_args() {
        let (system, user) =
            build_prompt("compress a.mp4", &registry(), &PromptOverrides::default());
        assert_eq!(user, "compress a.mp4");
        // the real tool is listed with required + optional (+ help)
        assert!(system.contains("  - compress_video: Compress a video."));
        assert!(system.contains("args: inputs (required), quality (optional, small_file|balanced)"));
        // system + internal tools are excluded
        assert!(!system.contains("- clarify"));
        assert!(!system.contains("- run_batch"));
        // defaults applied
        assert!(system.contains("You are a command planner"));
        assert!(system.contains("Examples:"));
    }

    #[test]
    fn overrides_replace_header_and_examples() {
        let overrides = PromptOverrides {
            system_header: Some("CUSTOM HEADER\n".to_string()),
            examples_block: Some("\nMY EXAMPLES".to_string()),
            ..Default::default()
        };
        let (system, _) = build_prompt("x", &registry(), &overrides);
        assert!(system.starts_with("CUSTOM HEADER\n"));
        assert!(system.ends_with("MY EXAMPLES"));
        assert!(!system.contains("You are a command planner")); // default header gone
    }

    #[test]
    fn render_examples_matches_python_shape() {
        let examples = vec![PromptExample {
            request: "compress a.mp4".into(),
            output: Some(
                serde_json::json!({"plan": [{"tool": "compress_video", "args": {"inputs": ["a.mp4"]}}]}),
            ),
        }];
        let block = render_examples(&examples).unwrap();
        assert!(block.starts_with("Examples:\n  request: \"compress a.mp4\"\n  output:  {"));
        // Must match Python `json.dumps(sep=(', ', ': '))`: spaced separators AND insertion-order
        // keys (tool before args) — the format the model was fine-tuned on.
        assert!(
            block.contains(
                "{\"plan\": [{\"tool\": \"compress_video\", \"args\": {\"inputs\": [\"a.mp4\"]}}]}"
            ),
            "example JSON must match python's spaced, tool-first shape:\n{block}"
        );
    }

    #[test]
    fn loads_real_ffmpeg_prompt_yaml() {
        let path = Path::new(env!("CARGO_MANIFEST_DIR")).join("../../../skills/ffmpeg/prompt.yaml");
        let overrides = load_prompt_yaml(&path);
        assert!(overrides.system_header.is_some());
        assert!(overrides
            .examples_block
            .as_deref()
            .unwrap()
            .contains("Examples:"));
    }
}
